optuna/importance/_mean_decrease_impurity.py

```diff
@@ -0,0 +1,136 @@
+            The maximum depth of each tree in the random forest.
+        random_seed:
+            Seed for the random forest.
+    """
+
+    def __init__(
+        self, n_estimators: int = 16, max_depth: int = 64, random_state: Optional[int] = None
+    ) -> None:
+        _check_sklearn_availability()
+
+        self._forest = RandomForestRegressor(
+            n_estimators=n_estimators,
+            max_depth=max_depth,
+            min_samples_split=2,
+            min_samples_leaf=1,
+            random_state=random_state,
+        )
+
+    def evaluate(self, study: Study, params: Optional[List[str]]) -> Dict[str, float]:
+        distributions = _get_distributions(study, params)
+        params_data, values_data = _get_study_data(study, distributions)
+
+        if params_data.size == 0:  # `params` were given but as an empty list.
+            return OrderedDict()
+
+        params_data, cols_to_raw_cols = _encode_categorical(params_data, distributions.values())
+
+        forest = self._forest
+        forest.fit(params_data, values_data)
+        feature_importances = forest.feature_importances_
+        feature_importances_reduced = numpy.zeros(len(distributions))
+        numpy.add.at(feature_importances_reduced, cols_to_raw_cols, feature_importances)
+
+        param_importances = OrderedDict()
+        param_names = list(distributions.keys())
+        for i in feature_importances_reduced.argsort()[::-1]:
+            param_importances[param_names[i]] = feature_importances_reduced[i].item()
+
+        return param_importances
+
+
+def _encode_categorical(
+    params_data: numpy.ndarray, distributions: ValuesView[BaseDistribution],
+) -> Tuple[numpy.ndarray, numpy.ndarray]:
+    # Transform the `params_data` matrix by expanding categorical integer-valued columns to one-hot
+    # encoding matrices. Note that the resulting matrix can be sparse and potetially very big.
+
+    numerical_cols = []
+    categorical_cols = []
+    categories = []
+
+    for col, distribution in enumerate(distributions):
+        if isinstance(distribution, CategoricalDistribution):
+            categories.append(numpy.arange(len(distribution.choices)))
+            categorical_cols.append(col)
+        else:
+            numerical_cols.append(col)
+    assert col == params_data.shape[1] - 1
+
+    col_transformer = ColumnTransformer(
+        [("_categorical", OneHotEncoder(categories=categories, sparse=False), categorical_cols)],
+        remainder="passthrough",
+    )
+    # All categorical one-hot columns are placed before the numerical columns in
+    # `ColumnTransformer.fit_transform`.
+    params_data = col_transformer.fit_transform(params_data)
+
+    # `cols_to_raw_cols["column index in transformed matrix"]
+    #     == "column index in original matrix"`
+    cols_to_raw_cols = numpy.empty((params_data.shape[1],), dtype=numpy.int32)
+
+    i = 0
+    for categorical_col in categorical_cols:
+        for _ in range(len(categories[i])):
```

Comment:
If we add one more categorical hyperparameter (e.g., `x7 = trial.suggest_categorical("x7", [1.0, 1.1, 1.2])`) to the objective function of `test_get_param_importances`, the test failed here.

I guess the end of the inner loop is supposed to be fixed during the iteration.
```suggestion
    for j, categorical_col in enumerate(categorical_cols):
        for _ in range(len(categories[j])):
```

Reply:
You're absolutely right. What do you think of using `zip` instead?

Reply:
Fixed it in https://github.com/optuna/optuna/pull/1253/commits/1aa2b4385e5e84e800f7d34b39e551c68b316ea2. The importance assessment with the PyTorch example with categorical was largely the same after the change.

Reply:
> You're absolutely right. What do you think of using zip instead?

It's much better than my proposal. Thanks!